napalm_ng/base.py

```diff
@@ -0,0 +1,217 @@
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations under
+# the License.
+
+# Python3 support
+from __future__ import print_function
+from __future__ import unicode_literals
+
+# stdlib
+import sys
+
+# local modules
+from napalm_ng import exceptions
+
+
+class BaseNetworkDriver(object):
+
+    def __init__(self, hostname, username, password, **kwargs):
+        """
+        This is the base class you have to inherit from when writing your own Network Driver to
+        manage any device. You will, in addition, have to override most of the methods specified on
+        this class. Make sure you follow the guidelines for every method and that you return the
+        correct data.
+
+        Args:
+
+            hostname (str): IP or FQDN of the device you want to connect to.
+            username (str): Username you want to use
+            password (str): Password
+            kwargs: Optional arguments that might tweak certain behaviors (see docs)
+        """
+        raise NotImplementedError
+
+    def __enter__(self):
+        """ Do not implement this. """
+        try:
+            self.open()
+        except Exception:
+            exc_info = sys.exc_info()
+            self.__raise_clean_exception(exc_info[0], exc_info[1], exc_info[2])
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        """ Do not implement this. """
+        self.close()
+        if exc_type is not None:
+            self.__raise_clean_exception(exc_type, exc_value, exc_traceback)
+
+    @staticmethod
+    def __raise_clean_exception(exc_type, exc_value, exc_traceback):
+        """
+        This method is going to check if the exception exc_type is part of the builtins exceptions
+        or part of the napalm exceptions. If it is not, it will print a message on the screen
+        giving instructions to fill a bug.
+
+        Finally it will raise the original exception.
+
+        Args:
+            exc_type: Exception class.
+            exc_value: Exception object.
+            exc_traceback: Traceback.
+
+        Raises:
+            Exception: The original exception
+        """
+        if (exc_type.__name__ not in dir(exceptions) and
+                exc_type.__name__ not in __builtins__.keys()):
+            epilog = ("NAPALM didn't catch this exception. Please, fill a bugfix on "
+                      "https://github.com/napalm-automation/napalm/issues\n"
+                      "Don't forget to include this traceback.")
+            print(epilog)
+        # Traceback should already be attached to exception; no need to re-attach
+        raise exc_value
+
+    def open(self):
+        """
+        Opens a connection to the device.
+
+        Raises:
+            (we should define here which exception we want to raise and in which cases)
```

Comment:
Totally agree on this.
I see an exceptions module below. Wouldn't we use the napalm-base exceptions?
We have this epic: https://github.com/napalm-automation/napalm-base/issues/218 which is supposed to track some more exceptions to be introduced.